### ISSUE TITLE
grad of output with respect to inputs  functions on cpu but not gpu

### ISSUE THREAD

**Issue by jkr26** [pytorch/pytorch#7425]
If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description
Taking gradient of outputs with respect to inputs (for e.g. visualizing where a CNN is 'looking' while classifying an image) functions on the cpu but not after the model and inputs are pushed to the gpu.

## Code example

```
import torch.nn as nn
import torch
from torch.autograd import grad

class testModule(nn.Module):
    def __init__(self):
        super(testModule, self).__init__()
        self.lin = nn.Linear(10, 1)

    def forward(self, x):
        return self.lin(x)

def test_cpu():
    t = torch.ones([1, 10], requires_grad=True)
    mod = testModule()
    output = mod(t)
    output[0].backward()
    test = t.grad
    return test

def test_gpu():
    mod = testModule().cuda()
    t = torch.ones([1, 10], requires_grad=True).cuda()
    output = mod(t)
    output[0].backward()
    test = t.grad
    return test

print(test_cpu())

print(test_gpu())
```

## System Info
- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): conda
- Build command you used (if compiling from source):
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 17.10
GCC version: (Homebrew gcc 5.5.0_4) 5.5.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 390.48
cuDNN version: 7.0.5

Versions of relevant libraries:
[pip3] numpy (1.14.2)
[pip3] numpydoc (0.7.0)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.1)
[conda] cuda90                    1.0                  h6433d27_0    pytorch
[conda] pytorch                   0.4.0           py36_cuda9.0.176_cudnn7.1.2_1  [cuda90]  pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch

**Comment by soumith** [comment]
Your code had a bug, I fixed it for you:

```
def test_gpu():
    mod = testModule().cuda()
    t = torch.ones([1, 10], requires_grad=True, device="cuda:0")
    output = mod(t)
    output[0].backward()
    test = t.grad
```

**Comment by donydchen** [comment]
Hi @soumith , thanks for your solution. Using your amended code fix the bug.

But I'm curious about the reason. The main difference between your code and the original post's is the way of tensor initialization. It looks like both ways are normal and correct ways to create tensor on CUDA device. 

So what really causes the difference? Could you please kindly enlighten me? Thanks.
